Rework builder module to get rid of all `HashMap<'static str, T>`

GnomedDev opened the pull request:
Swaps all the `HashMap<'static str, T>` for concrete rust structs and make them `derive(serde::Serialize)`, this allows for possible const builders, gets rid of a ton of heap allocations, and cleans up code massively. Nested builders are also serialised while building the top level one, meaning that all serialisation is done in one step. This also actually removes all usage of `json::to_value` outside one usage in a test, so I've `cfg(test)` gated that (and it shows that we no longer need Value, at least for serialisation).

Since I basically did a once over the whole builder module, some arguments have been changed to be more consistent or efficient, the builders internally store the `Type` enums instead of the raw number, and some int sizes have been reduced.

This does make all builder fields private, although a user can still pass custom JSON to the API via the raw HTTP functions.

GnomedDev committed:
De-hashmap EditVoiceState

GnomedDev committed:
Fix doc tests

GnomedDev committed:
Fmt

GnomedDev committed:
De-hashmap EditThread

GnomedDev committed:
De-hashmap EditRole

GnomedDev committed:
Fix EditRole

mkrasnitski commented:
Are we sure this is something we want to do? Any new fields added by discord will require breaking changes unless we make all builders `#[non_exhaustive]`. An alternative is to just make the hashmap private, then the methods on the builder are all that are exposed.

arqunis commented:
@mkrasnitski This is not an issue if the fields of the builders are private. Rust allows you to construct `struct`s only if all of their fields are `pub`lic. If one of them is private, users of the crate cannot instantiate the `struct`. `#[non_exhaustive]` is necessary when you want to impose this restriction for `struct`s that *do* have all fields `pub`lic, but is not necessarily open for direct construction, besides the crate where it is defined in, because you may want to add new fields in the future.

GnomedDev committed:
Swap taking Vec<u8> to taking impl serialize in http methods

Also fixes the tests

GnomedDev committed:
I love accidentally commiting dev stuff

GnomedDev committed:
De-hashmap AddMember

GnomedDev committed:
De-hashmap CreateAllowedMentions

GnomedDev committed:
De-hashmap Create Application Command Permission builders

GnomedDev committed:
De-hashmap CreateChannel

GnomedDev committed:
De-hashmap Component builders

GnomedDev committed:
De-hashmap CreateEmbed and similar

GnomedDev committed:
De-hashmap EditWebhookMessage

GnomedDev committed:
De-hashmap CreateMessage

GnomedDev committed:
De-hashmap CreateInteractionFollowup

GnomedDev committed:
De-hashmap CreateInvite

GnomedDev committed:
De-hashmap CreateScheduledEvent

GnomedDev committed:
De-hashmap Stage Channel Instance builders

GnomedDev committed:
De-hashmap Sticker builders

mkrasnitski commented:
Due to a limitation of `async fn`, having `'static` lifetimes in the params for `Http::create_sticker` causes incorrect lifetime unification. I think a solution is to either annotate the `&self` and `audit_log_reason` params with `'a`, or explicitly desugar the return type as `impl Future<Output = Result<Sticker>>> + 'a`, making sure to add the lifetime bound at the end there. Let me know which one ends up working.

The relevant issue is rust-lang/rust #63033/ (not linked as to not create a cross-reference). Not sure why only the MSRV build ran into this issue, as I think this fails to compile even on 1.61.

GnomedDev committed:
Fix clippy on CreateSticker

GnomedDev committed:
De-hashmap CreateThread

GnomedDev committed:
Make CreateSticker::build public

GnomedDev committed:
Try being explicit with lifetimes in create_sticker to support MSRV

GnomedDev committed:
Format

GnomedDev committed:
De-hashmap Application Commmand builders

GnomedDev committed:
De-hashmap CreateInteractionResponse and similar

GnomedDev committed:
De-hashmap EditChannel

GnomedDev committed:
De-hashmap EditGuildWelcomeScreen

GnomedDev committed:
De-hashmap EditGuildWidget

GnomedDev committed:
De-hashmap EditGuild

GnomedDev committed:
De-hashmap EditInteractionResponse

GnomedDev committed:
De-hashmap EditMember

GnomedDev committed:
Temp: add allows for unused functions

GnomedDev committed:
De-hashmap EditMessage

GnomedDev committed:
Cleanup ChannelId::create_permission

This is possible due to the From impl added previously in the CreateChannel rework

GnomedDev committed:
De-hashmap EditProfile

GnomedDev committed:
Fix Option<Option> fields

GnomedDev committed:
De-hashmap EditScheduledEvent

GnomedDev committed:
De-hashmap ExecuteWebhook

Also removes unneeded CONTENT_TYPE setting, as this is done inside Request::build anyway if content is passed (which it is)

GnomedDev committed:
Successfully compiling is nice

GnomedDev committed:
Fix tests

GnomedDev committed:
Fix feature build and MSRV
